File: swift/blackjack.swift
import Foundation

enum BlackjackValue {
    case single(Int)
    case double(Int, Int)

    var value: (Int, Int) {
        switch self {
        case .single(let value):
            return (value, value)
        case .double(let value1, let value2):
            return (value1, value2)
        }
    }
}

struct Card {
    let repr: String
    let value: BlackjackValue 
}

struct Deck {
    var cards: [Card] = []

    mutating func shuffle() {
        self.cards = self.cards.shuffled()
    }
    
    mutating func reset() {
        self.cards = []
        for value in 2...14 {
            for suit in ["♥", "♠", "♦", "♣"] {
                var newCard: Card

                switch value {
                case 11: newCard = Card(repr: "J\(suit)", value: BlackjackValue.single(10))
                case 12: newCard = Card(repr: "Q\(suit)", value: BlackjackValue.single(10))
                case 13: newCard = Card(repr: "K\(suit)", value: BlackjackValue.single(10))
                case 14: newCard = Card(repr: "A\(suit)", value: BlackjackValue.double(1, 11))
                default: newCard = Card(repr: "\(value)\(suit)", value: BlackjackValue.single(value))
                }

                self.cards.append(newCard)
            }
        }

        self.shuffle()
    }
    
    mutating func draw() -> Card? {
        guard self.cards.count > 0 else { return nil }
        return self.cards.removeFirst()
    }
}

struct BlackjackHand {
    var cards: [Card] = []
    var total: (Int, Int) = (0, 0)

    func status() {
        print(self.cards)
        print("Total cards: \(self.total)")
    }

    mutating func hit(deck: inout Deck) {
        guard deck.cards.count > 0 else { return }
        let nextCard = deck.draw()!
        let (value1, value2) = nextCard.value.value
        self.total.0 += value1
        self.total.1 += value2
        self.cards.append(nextCard)
    }
}

var deck = Deck()
deck.reset()

var hand = BlackjackHand()
hand.status()
hand.hit(deck: &deck)
hand.hit(deck: &deck)
hand.hit(deck: &deck)

hand.status()

File: nim/blackjack.nim
import std/[algorithm, random, strformat, strutils]

randomize()

type BlackjackValue = array[2, int]

type Card = object
  displayString: string
  value: BlackjackValue

type BlackjackDeck = object
  startingHandSize: int
  cards: seq[Card]
    
type BlackjackHand = object
  cards: seq[Card]
  total: BlackjackValue

proc status(hand: BlackjackHand) =
  for card in hand.cards:
    write(stdout, &"{card.displayString} ")
  write(stdout, "\n")
  flushFile(stdout)
  echo &"{hand.total[0]} {hand.total[1]}"

proc hit(hand: var BlackjackHand, card: Card) =
  hand.cards.add(card)
  hand.total[0] += card.value[0]
  hand.total[1] += card.value[1]

proc reset(deck: var BlackjackDeck) =
  deck.cards = @[]
  for chars in product([@["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"], @["♥", "♠", "♦", "♣"]]):
    var value =
      case chars[0]:
        of "A":
          [1, 11]
        of "10", "J", "Q", "K":
          [10, 10]
        else:
          [parseInt(chars[0]), parseInt(chars[0])]
    deck.cards.add(Card(displayString: chars.join(), value: value))
  deck.cards.shuffle()

proc hit(deck: var BlackjackDeck): Card =
  deck.cards.pop()
  
proc main() =
  var playerHand = BlackjackHand(cards: @[], total: [0, 0])
  var deck = BlackjackDeck(cards: @[], startingHandSize: 2)
  deck.reset()
  while playerHand.total[0] < 21:
    playerHand.hit(deck.hit())
    playerHand.status()

  echo "You've busted!"


main()
